File: java/146. LRU Cache.java
// hashtable with double linked list solution
class ListNode {
    int key;
    int val;
    ListNode prev;
    ListNode next;
    public ListNode (int key, int val, ListNode prev, ListNode next) {
        this.key = key;
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}

class LRUCache {
    private HashMap<Integer, ListNode> map = new HashMap<>();
    private int capacity;
    private ListNode head;
    private ListNode tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;

        this.head = new ListNode(-1, -1, null, null);
        this.tail = new ListNode(-1, -1, null, null);

        this.head.next = this.tail; // head -> tail
        this.tail.prev = this.head; // head <- tail
    }
    
    public int get(int key) {
        ListNode node = this.map.get(key); // get node by key
        if (node == null) return -1; // not found

        // update node position in double linked list
        this.remove(node);
        this.add(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        ListNode node;
        if (this.map.containsKey(key)) {
            // node has already existed in linked list
            node = this.map.get(key);
            this.remove(node);
            node.val = value;
        } else {
            // create new node
            node = new ListNode(key, value, null, null);
        }

        // update node position in double linked list
        this.add(node);
        this.map.put(key, node);

        // if map size exceed capacity, remove the least recently used node (right of the head)
        if(this.map.size() > this.capacity) {
            ListNode lruNode = this.head.next;
            this.remove(lruNode);
            this.map.remove(lruNode.key);
        }
    }

    public void add (ListNode node) {
        ListNode tailPrev = this.tail.prev;
        tailPrev.next = node; // tailPrev -> node
        node.next = this.tail; // node -> tail
        this.tail.prev = node; // node <- tail
        node.prev = tailPrev; // tailPrev <- node
    }

    public void remove (ListNode node) {
        ListNode nodePrev = node.prev;
        ListNode nodeNext = node.next;
        nodePrev.next = nodeNext;
        nodeNext.prev = nodePrev;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: golang/problems/medium/146. LRU Cache.go
package medium

import (
	"log"
)

type DoublePtrNode struct {
	Key  int
	Val  int
	Next *DoublePtrNode
	Prev *DoublePtrNode
}

type LRUCache struct {
	hashMap  map[int]*DoublePtrNode
	head     *DoublePtrNode
	tail     *DoublePtrNode
	capacity int
}

func LRUCacheConstructor(capacity int) LRUCache {
	head := &DoublePtrNode{}
	tail := &DoublePtrNode{}

	head.Next = tail
	tail.Prev = head

	return LRUCache{
		hashMap:  make(map[int]*DoublePtrNode, capacity),
		head:     head,
		tail:     tail,
		capacity: capacity,
	}
}

func (this *LRUCache) Get(key int) int {
	node, found := this.hashMap[key]
	if !found {
		return -1
	}

	if node.Prev != this.head {
		node.Prev.Next = node.Next
		node.Next.Prev = node.Prev

		next := this.head.Next
		this.head.Next = node
		node.Prev = this.head
		node.Next = next
		next.Prev = node
	}
	return node.Val
}

func (this *LRUCache) Put(key int, value int) {
	if node, found := this.hashMap[key]; found {
		node.Val = value
		this.Get(key)
		return
	}

	if len(this.hashMap) == this.capacity {
		lruNode := this.tail.Prev
		delete(this.hashMap, lruNode.Key)
		this.tail.Prev = lruNode.Prev
		lruNode.Prev.Next = this.tail
	}

	node := &DoublePtrNode{Key: key, Val: value}
	this.hashMap[key] = node
	next := this.head.Next
	this.head.Next = node
	node.Prev = this.head
	node.Next = next
	next.Prev = node
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

func Test_LRUCache() {
	//obj := LRUCacheConstructor(2)
	//
	//obj.Put(1, 1)
	//obj.Put(2, 2)
	//log.Println(obj.Get(1))
	//obj.Put(3, 3)
	//log.Println(obj.Get(2))
	//obj.Put(4, 4)
	//log.Println(obj.Get(1))
	//log.Println(obj.Get(3))
	//log.Println(obj.Get(4))

	obj2 := LRUCacheConstructor(2)

	obj2.Put(1, 0)
	obj2.Put(2, 2)
	log.Println(obj2.Get(1))
	obj2.Put(3, 3)
	log.Println(obj2.Get(2))
	obj2.Put(4, 4)
	log.Println(obj2.Get(1))
	log.Println(obj2.Get(3))
	log.Println(obj2.Get(4))
}
